MKLDNN-DO Subgraph Optimization

sreekanth-yalachigere opened the pull request:
This optimization creates a subgraph of mkldnn operators and uses full potential of mkldnn by using block propagation and improves performance significantly.
For example, in Resnet50, currently we have 109 data reorders.

```
onnxruntime_data(nchw) ->reorder(nchw8c)->conv(nchw8c) -> reorder(nchw);
onnxruntime_data(nchw) ->batchnorm(nchw);
onnxruntime_data(nchw) ->Relu(nchw);
onnxruntime_data(nchw) ->reorder(nchw8c)->pool(nchw8c) -> reorder(nchw);
onnxruntime_data(nchw) ->reorder(nchw8c)->conv(nchw8c) -> reorder(nchw);

```
With this optimization, we will have only one data reorder and we propage blocked mkldnn memory all the way to the end of subgraph.

`onnxruntime_data(nchw) -> conv(nchw8c) -> batchnorm-relu(nchw8c)->pool(nchw8c)->conv(nchw8c)... reorder(nchw)`

Subgraph optimization can be enabled by setting the following environment variable.
`ORT_MKLDNN_SUBGRAPH=1`

jywu-msft commented:
/azp run

pranavsharma commented on onnxruntime/core/providers/mkldnn/subgraph/mkldnn_custom_op.h:
```diff
@@ -0,0 +1,283 @@
+      mkldnn::memory::dims src_dims(x_shape.GetDims().begin(), x_shape.GetDims().end());
+      std::string key;
+      key.reserve(128);
+      key = subgraph_key_;
```
Can AddDimsToKey be written such that it accepts a const string and returns the newly formed key? Moreover, we should rename this to GenerateKey() as addition is just one way to generate the key.

sreekanth-yalachigere replied:
@pranavsharma  AddDimsToKey is widely used in Vanilla mkldnn operators. Should I go ahead and modify it?

pranavsharma commented on onnxruntime/core/providers/mkldnn/subgraph/mkldnn_custom_op.h:
```diff
@@ -0,0 +1,283 @@
+    mkldnn::memory::format source_format = mkldnn::memory::format::any;  // ONNXRuntime format
+    for (auto& kernel : context_.kernels) {
+      Status status = kernel->CreatePrimitives(input_tensors, cpu_engine_, context_.net, source_format);
+      if (status.IsOK())
```
Looks like Initialize will fail silently. Can you make it return a status code and throw in the constructor if a failure is encountered?

pranavsharma commented on onnxruntime/core/providers/mkldnn/mkldnn_execution_provider.cc:
```diff
@@ -107,6 +111,276 @@ std::shared_ptr<KernelRegistry> GetMklDnnKernelRegistry() {
+      if (attributes.size() > 0) {
+        int index = static_cast<int>(sub_var.subgraph_ptr->mklnodes.size());
+
+        for (auto attIt = attributes.begin(); attIt != attributes.end(); ++attIt) {
```
naming convention. should be att_it

pranavsharma commented on onnxruntime/core/providers/mkldnn/mkldnn_execution_provider.cc:
```diff
@@ -107,6 +111,276 @@ std::shared_ptr<KernelRegistry> GetMklDnnKernelRegistry() {
+  return use_subgraph;
+}
+
+std::vector<std::unique_ptr<ComputeCapability>> MKLDNNExecutionProvider::GetCapability(
```
Still reading this method. It is really long and difficult to read. Can it be split?

sreekanth-yalachigere replied:
Extracted new method to create MklDnnNode.

```
// Create MklDnn node, update inputs, outputs and parent nodes
// collect attribtes
void CreateMetaDef(SubgraphVariables& sub_var, const NodeAttributes& subgraph_attributes,
                     std::vector<std::unique_ptr<ComputeCapability>>& result) const;
```

jywu-msft commented on onnxruntime/core/providers/mkldnn/mkldnn_execution_provider.cc:
```diff
@@ -80,19 +84,19 @@ class ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kMklDnnExecutionProvider, kOnnxDomai
-    BuildKernelCreateInfo<ONNX_OPERATOR_VERSIONED_TYPED_KERNEL_CLASS_NAME(kMklDnnExecutionProvider, kOnnxDomain, 8, 8, float, MaxPool)>,
-    BuildKernelCreateInfo<ONNX_OPERATOR_VERSIONED_TYPED_KERNEL_CLASS_NAME(kMklDnnExecutionProvider, kOnnxDomain, 1, 8, float, GlobalMaxPool)>,
-    BuildKernelCreateInfo<ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kMklDnnExecutionProvider, kOnnxDomain, 1, float, LRN)>,
+      BuildKernelCreateInfo<ONNX_OPERATOR_KERNEL_CLASS_NAME(kMklDnnExecutionProvider, kOnnxDomain, 1, Conv)>,
```
why did the formatting here change?

jywu-msft commented on onnxruntime/core/providers/mkldnn/mkldnn_execution_provider.cc:
```diff
@@ -107,6 +111,276 @@ std::shared_ptr<KernelRegistry> GetMklDnnKernelRegistry() {
+
+  // temporary switch to toggle between mkldnn-vanilla and mkldnn-subgraph implementation using
+  // ORT_MKLDNN_SUBGRAPH environment variable
+  if (RunSubgraph(graph_viewer, kernel_registries, result) == false){
```
agreed, can use the base class GetCapability() instead of duplicating its code in RunSubGraph()

jywu-msft commented on onnxruntime/core/providers/mkldnn/mkldnn_execution_provider.cc:
```diff
@@ -107,6 +111,276 @@ std::shared_ptr<KernelRegistry> GetMklDnnKernelRegistry() {
+  }
+
+  // use sub-graph implementation
+  std::vector<std::string> mkl_ops = {"Conv", "BatchNormalization", "Relu", "Sum",
```
gets lost if kept as a local variable here. should be part of the class definition? 
and it's probably better to use a set instead of vector.

jywu-msft commented on onnxruntime/core/providers/mkldnn/subgraph/subgraph.h:
```diff
@@ -0,0 +1,68 @@
+
+namespace onnxruntime {
+
+struct MklNode {
```
agreed. I think anything named mkl should be renamed mkldnn to avoid confusion.
this includes file names.
e.g. mkl_kernel.h , mklconv.h etc.

sreekanth-yalachigere replied:
Changed it to MklDnn

jywu-msft commented on onnxruntime/core/providers/mkldnn/subgraph/MklConv.h:
```diff
@@ -0,0 +1,588 @@
+// Copyright(C) 2019 Intel Corporation
```
if you used/modified code from existing onnxruntime, please ensure you carry over the microsoft license header as well (additional intel header can be below it)

this applies for all files. thanks.

jywu-msft commented on onnxruntime/core/providers/mkldnn/subgraph/MklSum.h:
```diff
@@ -0,0 +1,148 @@
+    int num_inputs = mklnode_ptr_->num_inputs;
+    int input_index = mklnode_ptr_->input_start_index < 0 ? 0 : mklnode_ptr_->input_start_index;
+
+    std::vector<float> coeff;
```
where is coeff used

jywu-msft commented:
can you add some high level documentation, which describes the classes, abstractions and their relationships.
this would help maintenance of this code moving forward.

> This optimization creates a subgraph of mkldnn operators and uses full potential of mkldnn by using block propagation and improves performance significantly.
> For example, in Resnet50, currently we have 109 data reorders.
> 
> ```
> onnxruntime_data(nchw) ->reorder(nchw8c)->conv(nchw8c) -> reorder(nchw);
> onnxruntime_data(nchw) ->batchnorm(nchw);
> onnxruntime_data(nchw) ->Relu(nchw);
> onnxruntime_data(nchw) ->reorder(nchw8c)->pool(nchw8c) -> reorder(nchw);
> onnxruntime_data(nchw) ->reorder(nchw8c)->conv(nchw8c) -> reorder(nchw);
> ```
> 
> With this optimization, we will have only one data reorder and we propage blocked mkldnn memory all the way to the end of subgraph.
> 
> `onnxruntime_data(nchw) -> conv(nchw8c) -> batchnorm-relu(nchw8c)->pool(nchw8c)->conv(nchw8c)... reorder(nchw)`
> 
> Subgraph optimization can be enabled by setting the following environment variable.
> `ORT_MKLDNN_SUBGRAPH=1`

jywu-msft commented on onnxruntime/core/providers/mkldnn/subgraph/subgraph.h:
```diff
@@ -0,0 +1,68 @@
+#include <map>
+
+namespace onnxruntime {
+
```
mkl_dnn namespace?

jywu-msft commented on onnxruntime/core/providers/mkldnn/subgraph/MklBatchNorm.h:
```diff
@@ -0,0 +1,308 @@
+                        MKLDNNExecutionProvider* provider,
+                        std::shared_ptr<MKLContext> mkl_context) : MklKernel(node, provider, mkl_context) {
+  }
+  void ReadAttributes(const std::unordered_map<std::string,
```
unordered_map<std::string, ONNX_NAMESPACE::AttributeProto> is  just NodeAttributes.
seems to be inconsistently used with your changes. 
please replace all occurrences

jywu-msft commented on onnxruntime/core/providers/mkldnn/subgraph/mkldnn_custom_op.h:
```diff
@@ -0,0 +1,283 @@
+        os << "Conv-" << mklnode.node_index << "-";
+        std::shared_ptr<MklConv<T>> kernel;
+        kernel.reset(new MklConv<T>(mklnode, params.provider, params.mkl_context));
+        kernel->ReadAttributes(params.attributes, os.str());
```
could populating the attributes be done as part of kernel construction? 
instead of having another public method that always gets called right after?
you can keep the method around as a private helper. but call it in the constructor?

jywu-msft commented on onnxruntime/core/providers/mkldnn/subgraph/mkldnn_custom_op.h:
```diff
@@ -0,0 +1,283 @@
+    }
+  }
+
+  void CreateKernels(const SubgraphParams& params) {
```
does this need to be public?

sreekanth-yalachigere replied:
made private.

jywu-msft commented on onnxruntime/core/providers/mkldnn/subgraph/mkldnn_custom_op.h:
```diff
@@ -0,0 +1,283 @@
+
+namespace {
+struct SubgraphParams {
+  std::unordered_map<std::string, ONNX_NAMESPACE::AttributeProto> attributes;
```
use NodeAttributes

jywu-msft commented on onnxruntime/core/providers/mkldnn/subgraph/mkldnn_custom_op.h:
```diff
@@ -0,0 +1,283 @@
+}  // namespace
+
+template <typename T>
+class MkldnnCustomOp {
```
MkldnnFuncKernel might be better name.

jywu-msft commented on onnxruntime/core/providers/mkldnn/subgraph/mkl_kernel.h:
```diff
@@ -0,0 +1,178 @@
+  }
+  virtual ~MklKernel(){};
+
+  virtual void ReadAttributes(const std::unordered_map<std::string,
```
is it possible to leverage OpNodeProtoHelper for reading attributes? 
then you wouldn't need to define all the Get*Attr() methods below.

sreekanth-yalachigere replied:
I looked into it and it's not straight forward. Can this be done after this PR?

jywu-msft commented on onnxruntime/core/providers/mkldnn/subgraph/subgraph.h:
```diff
@@ -0,0 +1,68 @@
+
+namespace onnxruntime {
+
+struct MklNode {
```
there's already onnxruntime representations of Node, subgraphs etc.
Were those too heavy to use? Would have been nice to leverage more of onnxruntime classes where possible,
rather than adding new ones to bridge onnxruntime and mkldnn library.

sreekanth-yalachigere replied:
MklDnnNode is totally different and I am using it for MklDnn IR of a subgraph.

yufenglee commented:
I tried this PR on a production model and got a crash. Here is the call stack:

#0  0x000000000047e487 in std::__find_if<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >, __gnu_cxx::__ops::_Iter_equals_val<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const> > (__first="Conv",
    __last=<error: Cannot access memory at address 0x8>, __pred=...) at /usr/include/c++/5/bits/stl_algo.h:118
#1  0x00000000004935ab in std::__find_if<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >, __gnu_cxx::__ops::_Iter_equals_val<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const> > (__pred=..., __last=...,
    __first=...) at /usr/include/c++/5/bits/stl_algo.h:162
#2  std::find<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > (__val=..., __last=..., __first=...) at /usr/include/c++/5/bits/stl_algo.h:3791
#3  onnxruntime::MKLDNNExecutionProvider::GetCapability (this=this@entry=0xe088f0, graph_viewer=..., kernel_registries=std::vector of length 1, capacity 1 = {...})
    at /home/yufeng/project/mkldnn/onnxruntime/onnxruntime/core/providers/mkldnn/mkldnn_execution_provider.cc:167
#4  0x0000000000788391 in onnxruntime::GraphPartitioner::Partition (this=this@entry=0x7fffffffbf50, graph=..., export_dll=false, func_mgr=...) at /home/yufeng/project/mkldnn/onnxruntime/onnxruntime/core/framework/graph_partitioner.cc:151
#5  0x00000000004663ee in onnxruntime::InferenceSession::TransformGraph (this=this@entry=0xe07150, graph=..., graph_transformer_mgr=..., providers=..., kernel_registry_manager=..., insert_cast_transformer=..., session_state=...)
    at /home/yufeng/project/mkldnn/onnxruntime/onnxruntime/core/session/inference_session.cc:301
#6  0x000000000046b7ce in onnxruntime::InferenceSession::Initialize (this=this@entry=0xe07150) at /home/yufeng/project/mkldnn/onnxruntime/onnxruntime/core/session/inference_session.cc:446
#7  0x0000000000450122 in OrtCreateSession (env=env@entry=0xe03ac0, model_path=0xcd9a50 "/home/yufeng/data/models/fastrcnn/perf/model.onnx", options=options@entry=0xe06cb0, out=out@entry=0x7fffffffd830)
    at /home/yufeng/project/mkldnn/onnxruntime/onnxruntime/core/session/onnxruntime_c_api.cc:390
#8  0x0000000000421903 in onnxruntime::SessionOptionsWrapper::OrtCreateSession (model_path=<optimized out>, this=<synthetic pointer>) at /home/yufeng/project/mkldnn/onnxruntime/include/onnxruntime/core/session/onnxruntime_cxx_api.h:127
#9  onnxruntime::perftest::OnnxRuntimeTestSession::OnnxRuntimeTestSession (this=0xdfe920, env=0xe03ac0, performance_test_config=..., m=0xdfdf20) at /home/yufeng/project/mkldnn/onnxruntime/onnxruntime/test/perftest/ort_test_session.cc:83
#10 0x0000000000423c41 in onnxruntime::perftest::CreateSession (test_model_info=0xdfdf20, performance_test_config_=..., env=0xe03ac0) at /home/yufeng/project/mkldnn/onnxruntime/onnxruntime/test/perftest/performance_runner.cc:188
#11 onnxruntime::perftest::PerformanceRunner::PerformanceRunner (this=0x7fffffffdf30, env=0xe03ac0, test_config=...) at /home/yufeng/project/mkldnn/onnxruntime/onnxruntime/test/perftest/performance_runner.cc:201
#12 0x0000000000420bbe in real_main (argc=<optimized out>, argv=<optimized out>, p_env=p_env@entry=0x7fffffffe160) at /home/yufeng/project/mkldnn/onnxruntime/onnxruntime/test/perftest/main.cc:33
#13 0x0000000000419615 in main (argc=<optimized out>, argv=<optimized out>) at /home/yufeng/project/mkldnn/onnxruntime/onnxruntime/test/perftest/main.cc:53

yufenglee commented on onnxruntime/core/providers/mkldnn/mkldnn_execution_provider.cc:
```diff
@@ -107,6 +111,276 @@ std::shared_ptr<KernelRegistry> GetMklDnnKernelRegistry() {
+  while (node_index < graph_viewer.MaxNodeIndex()) {
+    auto node = graph_viewer.GetNode(node_index);
+    std::vector<std::string>::iterator it = std::find(
+        mkl_ops.begin(), mkl_ops.end(), node->OpType());
```
the crash is because node is nullptr. Need to figure out why node is null.

sreekanth-yalachigere replied:
@yufenglee is this PR model available on ONNX Model Zoo?

yufenglee replied:
@sreekanth-yalachigere , no, it is a production model and I can't share it with you for confidentiality.

jywu-msft replied:
GetNode() may return nullptr if the node has been freed , which can happen if there are some transforms.
I don't think you can iterate through nodes by index from 0 to max.
Shouldn't you use GetNodesInTopologicalOrder() ?

sreekanth-yalachigere commented on onnxruntime/core/providers/mkldnn/subgraph/mkl_kernel.h:
```diff
@@ -0,0 +1,178 @@
+    ORT_UNUSED_PARAMETER(attributes_prefix);
+  }
+
+  virtual Status CreatePrimitives(const ONNXRunTimeTensor* input_tensors,
```
I think input_tensors is correct. We access shape, dim and data by an index. Can you please confirm?

sreekanth-yalachigere commented:
@pranavsharma @jywu-msft 
1. Changed Mkl to MklDnn 
2. Renamed files (mkl_* to mkldnn_*)
3. Moved code from mkldnn_kernel.h to new mkldnn_kernel.cc
4. Extracted method from GetCapability long code.

sreekanth-yalachigere commented:
Creating new PR as many files have modified after memcpy code merge.
